vslib/src/TrafficForwarder.cpp

```diff
@@ -0,0 +1,100 @@
+#include "TrafficForwarder.h"
+
+#include "swss/logger.h"
+
+#include <memory.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <net/if.h>
+#include <linux/if_tun.h>
+#include <sys/ioctl.h>
+#include <net/if_arp.h>
+#include <unistd.h>
+#include <net/ethernet.h>
+#include <arpa/inet.h>
+#include <linux/if_packet.h>
+#include <linux/if_ether.h>
+
+using namespace saivs;
+
+#define ETH_FRAME_BUFFER_SIZE (0x4000)
+#define CONTROL_MESSAGE_BUFFER_SIZE (0x1000)
+#define IEEE_8021Q_ETHER_TYPE (0x8100)
+#define MAC_ADDRESS_SIZE (6)
+#define VLAN_TAG_SIZE (4)
+
+void TrafficForwarder::add_vlan_tag(
+    _Inout_ unsigned char *buffer,
+    _Inout_ size_t &length,
+    _Inout_ struct msghdr &msg) const
+{
+    SWSS_LOG_ENTER();
+
+    struct cmsghdr *cmsg;
+
+    for (cmsg = CMSG_FIRSTHDR(&msg); cmsg != NULL; cmsg = CMSG_NXTHDR(&msg, cmsg))
+    {
+        if (cmsg->cmsg_level != SOL_PACKET || cmsg->cmsg_type != PACKET_AUXDATA)
+            continue;
+
+        struct tpacket_auxdata* aux = (struct tpacket_auxdata*)CMSG_DATA(cmsg);
+
+        if ((aux->tp_status & TP_STATUS_VLAN_VALID) &&
+                (aux->tp_status & TP_STATUS_VLAN_TPID_VALID))
+        {
+            SWSS_LOG_DEBUG("got vlan tci: 0x%x, vlanid: %d", aux->tp_vlan_tci, aux->tp_vlan_tci & 0xFFF);
+
+            // inject vlan tag into frame
+
+            // for overlapping buffers
+            memmove(buffer + 2 * MAC_ADDRESS_SIZE + VLAN_TAG_SIZE,
+                    buffer + 2 * MAC_ADDRESS_SIZE,
+                    length - (2 * MAC_ADDRESS_SIZE));
+
+            uint16_t tci = htons(aux->tp_vlan_tci);
+            uint16_t tpid = htons(IEEE_8021Q_ETHER_TYPE);
+
+            uint8_t* pvlan =  (uint8_t *)(buffer + 2 * MAC_ADDRESS_SIZE);
+            memcpy(pvlan, &tpid, sizeof(uint16_t));
+            memcpy(pvlan + sizeof(uint16_t), &tci, sizeof(uint16_t));
+
+            length += VLAN_TAG_SIZE;
+
+            break;
+        }
+    }
+}
+
+bool TrafficForwarder::send_to(
+    _In_ int fd,
+    _In_ const unsigned char *buffer,
+    _In_ size_t &length) const
```

Comment:
length is not modified here no need for &

Reply:
Done